### PR TITLE
fix: shouldn't pass alias through into subquery.

### LINKED ISSUE #4123: Bug in interpreting correctly parsed SQL with aliases
**Describe the bug**

Some statements that involve subqueries with an alias, are not evaluated correctly when the subquery contains both `group by` and `order by`.

**To Reproduce**

With `datafusion-cli`:

```console
❯ select a from (select a from (select 1 as a) as foo group by 1 order by a) as c;
SchemaError(FieldNotFound { field: Column { relation: Some("foo"), name: "a" }, valid_fields: Some([Column { relation: Some("c"), name: "a" }]) })
```

**Expected behavior**

```console
❯ select a from (select a from (select 1 as a) as foo group by 1 order by a) as c;
+---+
| a |
+---+
| 1 |
+---+
1 row in set. Query took 0.005 seconds.
```

(it works on postgres and sqlite)

**Additional context**

Same happens if there is an actual table called foo:

```console
❯ create table foo as select 1 as a;
0 rows in set. Query took 0.032 seconds.
❯ select a from (select a from foo group by 1 order by a) as c;
SchemaError(FieldNotFound { field: Column { relation: Some("foo"), name: "a" }, valid_fields: Some([Column { relation: Some("c"), name: "a" }]) })
```

Interestingly if I remove as c in the outer expression it works:

```console
❯ select a from (select a from (select 1 as a) as foo group by 1 order by a);
+---+
| a |
+---+
| 1 |
+---+
1 row in set. Query took 0.005 seconds.
```

It's also interesting that using positional column references in the order by works:

```console
❯ select a from (select a from (select 1 as a) as foo group by 1 order by 1) as c;
+---+
| a |
+---+
| 1 |
+---+
1 row in set. Query took 0.005 seconds.
```

Furthermore, the bug reproduces only if `group by a` and `order by a` are *both* present:

```console
❯ select a from (select a from (select 1 as a) as foo group by a) as c;
+---+
| a |
+---+
| 1 |
+---+
1 row in set. Query took 0.005 seconds.
❯ select a from (select a from (select 1 as a) as foo order by a) as c;
+---+
| a |
+---+
| 1 |
+---+
1 row in set. Query took 0.004 seconds.
```

---

It appears that the alias applied to a subquery somehow affects the field references inside the subquery. For example this this should return an error because `c` is not defined inside the `select a from foo group by 1 order by c.a` query:

```console
❯ select a from (select a from foo group by 1 order by c.a) as c;
+---+
| a |
+---+
| 1 |
+---+
1 row in set. Query took 0.024 seconds.
```

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #4123.

# Rationale for this change

# What changes are included in this PR?

This PR has fixed a bug(We can see bug in UT and test case, but no one noticed it).

Original code passed the outside `Alias` into `Subquery` inside, it's wrong.

```sql
create table foo as select 1 as a;
select a from (select a from foo group by 1 order by a) as c;

Projection: c.a
  Projection: c.a, alias=c
    Projection: c.a
      Sort: foo.a ASC NULLS LAST
        Projection: foo.a, alias=c        <------- wrong, outside pass into subquery inside
          Aggregate: groupBy=[[foo.a]], aggr=[[]]
            TableScan: foo
```

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -778,7 +778,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 let normalized_alias = alias.as_ref().map(|a| normalize_ident(&a.name));
                 let logical_plan = self.query_to_plan_with_alias(
                     *subquery,
-                    normalized_alias.clone(),
+                    None,
```
Comment (jackwener) [inline datafusion/sql/src/planner.rs:781]: We shouldn't pass the `alias` into subquery inside.

File: datafusion/core/tests/sql/explain_analyze.rs [hunk 2]
```diff
@@ -746,7 +746,7 @@ async fn test_physical_plan_display_indent_multi_children() {
         "          ProjectionExec: expr=[c2@0 as c2]",
         "            ProjectionExec: expr=[c1@0 as c2]",
         "              RepartitionExec: partitioning=RoundRobinBatch(9000)",
-        "                CsvExec: files=[ARROW_TEST_DATA/csv/aggregate_test_100.csv], has_header=true, limit=None, projection=[c1]",
+        "                CsvExec: files=[ARROW_TEST_DATA/csv/aggregate_test_100.csv], has_header=true, limit=None, projection=[c1, c2]",
```
Comment (alamb) [inline datafusion/core/tests/sql/explain_analyze.rs:749]: This doesn't seem right to me.

The query is only selecting `c1` from the table -- so why is this scan now selecting c1 *and* c2?

```
    let sql = "SELECT c1 \
               FROM (select c1 from aggregate_test_100) AS a \
               JOIN\
               (select c1 as c2 from aggregate_test_100) AS b \
               ON c1=c2\
               ";
```

File: benchmarks/expected-plans/q13.txt [hunk 3]
```diff
@@ -2,8 +2,8 @@ Sort: custdist DESC NULLS FIRST, c_orders.c_count DESC NULLS FIRST
   Projection: c_orders.c_count, COUNT(UInt8(1)) AS custdist
     Aggregate: groupBy=[[c_orders.c_count]], aggr=[[COUNT(UInt8(1))]]
       Projection: c_orders.COUNT(orders.o_orderkey) AS c_count, alias=c_orders
-        Projection: c_orders.COUNT(orders.o_orderkey), alias=c_orders
-          Projection: COUNT(orders.o_orderkey), alias=c_orders
+        Projection: COUNT(orders.o_orderkey), alias=c_orders
+          Projection: COUNT(orders.o_orderkey)
```
Comment (alamb) [inline benchmarks/expected-plans/q13.txt:6]: This redundant projection seems unfortunate (though not caused by this PR)

File: datafusion/core/tests/sql/joins.rs [hunk 4]
```diff
@@ -1521,8 +1521,8 @@ async fn reduce_left_join_3() -> Result<()> {
         "Explain [plan_type:Utf8, plan:Utf8]",
         "  Projection: t3.t1_id, t3.t1_name, t3.t1_int, t2.t2_id, t2.t2_name, t2.t2_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
         "    Left Join: t3.t1_int = t2.t2_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
-        "      Projection: t3.t1_id, t3.t1_name, t3.t1_int, alias=t3 [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N]",
-        "        Projection: t1.t1_id, t1.t1_name, t1.t1_int, alias=t3 [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N]",
```
Comment (alamb) [inline datafusion/core/tests/sql/joins.rs:1525]: it sure looks better to have not have this double alias

File: datafusion/sql/src/planner.rs [hunk 5]
```diff
@@ -3170,10 +3170,10 @@ mod tests {
                      ) AS a
                    ) AS b";
         let expected = "Projection: b.fn2, b.last_name\
-                        \n  Projection: b.fn2, b.last_name, b.birth_date, alias=b\
-                        \n    Projection: a.fn1 AS fn2, a.last_name, a.birth_date, alias=b\
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:3174]: yeah looking at these old plans now this selecting a just to alias them as `b` is a little strange

**Review by alamb — APPROVED** [review]
I went through the plans carefully -- thank you @jackwener 

I can't quite figure out if this could cause an error / incorrect results but it certainly looks better

I am a little worried about the new column projection in datafusion/core/tests/sql/explain_analyze.rs but maybe that is a preexisting bug that is simply uncovered by this change.

**Comment by andygrove** [comment]
I took a first pass through this, and it seems to make sense, but I was confused by some of the expected results. I'll take another look tomorrow.

File: datafusion/core/tests/sql/explain_analyze.rs (hunk 2 above)
Comment (jackwener) [reply datafusion/core/tests/sql/explain_analyze.rs:749]: I think it should be caused by `project_pushdown rule`, I will look into it.

**Comment by jackwener** [comment]
@alamb @andygrove 
I have find the reason of `projection include [c1, c2]`.
It's because the `pushdown project` rule don't handle `alias`. In fact, when handle `projection with alias`, we need replace the `alias column` in `new_required_columns`.

For example:

```
project c1 as c2
    tablescan c1.
```

when we meet `project c1 as c2`, if `new_required_columns` include c2, we need replace it with c1.

**Comment by jackwener** [comment]
cc @mingmwang @liukun4515

**Comment by andygrove** [comment]
@jackwener Thanks for tracking down the bug and filing an issue.

**Review by andygrove — APPROVED** [review]

**Comment by liukun4515** [comment]
Thanks @jackwener

**Pull request merged**